### PR TITLE
Make primal stochastic trainer also typed

### LINKED ISSUE #2555: All linear trainers should be typed
There are still some binary classification trainers to (generalized) linear models. As done in  #2506, they all should be typed according to the types learned. They also should not do auto-calibration in API land.

We should also check if making other trainers typed is possible.

- EnsembleTrainer
- RegressionEnsembleTrainer
- MulticlassDataPartitionEnsembleTrainer
- MetaMulticlassTrainer
- Ova related things
- Stacking
- TreeEnsembleFeaturizer

### PR DESCRIPTION
**Pull request by wschin**
To fix #2555, we apply idea done in #2506 to another stochastic gradient trainer. Basically, to align this trainer and the whole SDCA family, `StochasticGradientDescentClassificationTrainer` got renamed to
```csharp
    public abstract class SgdBinaryTrainerBase<TModelParameters> :
        LinearTrainerBase<BinaryPredictionTransformer<TModelParameters>, TModelParameters>
        where TModelParameters : class, IPredictorProducing<float>
```
which holds common functions used in all derived classes.
The new class, `SgdBinaryTrainerBase`, branches out into three actual implementations:
- `LegacySgdBinaryTrainer` for command line and entry point BC.
- `SgdNonCalibratedBinaryTrainer` for training `LinearBinaryModelParameters` with any loss which implements `IClassificationLoss`.
- `SgdBinaryTrainer` for training `CalibratedModelParametersBase<LinearBinaryModelParameters, PlattCalibrator>`, which is essentially logistic regression and was the default before this PR.

### CODE REVIEW

File: src/Microsoft.ML.StandardLearners/StandardLearnersCatalog.cs [hunk 1]
```diff
@@ -30,34 +28,32 @@ public static class StandardLearnersCatalog
...
-            float l2Weight = SgdOptions.Defaults.L2Weight,
-            ISupportClassificationLossFactory loss = null)
+            int maxIterations = SgdBinaryTrainer.Options.Defaults.MaxIterations,
+            double initLearningRate = SgdBinaryTrainer.Options.Defaults.InitLearningRate,
+            float l2Weight = SgdBinaryTrainer.Options.Defaults.L2Weight)
         {
             Contracts.CheckValue(catalog, nameof(catalog));
             var env = CatalogUtils.GetEnvironment(catalog);
-            return new StochasticGradientDescentClassificationTrainer(env, labelColumn, featureColumn, weights, maxIterations, initLearningRate, l2Weight, loss);
+            return new SgdBinaryTrainer(env, labelColumn, featureColumn, weights, maxIterations, initLearningRate, l2Weight);
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.StandardLearners/StandardLearnersCatalog.cs]: >SgdBinaryTrainer [](start = 23, length = 16)

Why it become `SgdBinaryTrainer` instead of `StochasticGradientDescentClassificationTrainer`?
Method name is  `StochasticGradientDescent`

File: src/Microsoft.ML.StandardLearners/Standard/SdcaBinary.cs [hunk 2]
```diff
@@ -1716,18 +1715,12 @@ protected override IPredictorWithFeatureWeights<float> CreatePredictor(VBuffer<f
...
+    public abstract class SgdBinaryTrainerBase<TModelParameters> :
+        LinearTrainerBase<BinaryPredictionTransformer<TModelParameters>, TModelParameters>
+        where TModelParameters : class, IPredictorProducing<float>
     {
-        internal const string LoadNameValue = "BinarySGD";
-        internal const string UserNameValue = "Hogwild SGD (binary)";
-        internal const string ShortName = "HogwildSGD";
-
-        public sealed class Options : LearnerInputBaseWithWeight
+        public class SgdBinaryOptionsBase : LearnerInputBaseWithWeight
```
Comment (TomFinley) [inline src/Microsoft.ML.StandardLearners/Standard/SdcaBinary.cs]: >SgdBinaryOptionsBase [](start = 21, length = 20)

This is a nested class. Why is it prefixed with `SgdBinary`?

**Comment by TomFinley** [comment]
`OptionsBase` would be more consistent with what we seem to be doing, but I can just fix that fairly easily I think.

File: src/Microsoft.ML.StandardLearners/Standard/SdcaBinary.cs [hunk 3]
```diff
@@ -1817,75 +1805,62 @@ internal static class Defaults
...
-        protected override BinaryPredictionTransformer<TScalarPredictor> MakeTransformer(TScalarPredictor model, DataViewSchema trainSchema)
-            => new BinaryPredictionTransformer<TScalarPredictor>(Host, model, trainSchema, FeatureColumn.Name);
+        protected override BinaryPredictionTransformer<TModelParameters> MakeTransformer(TModelParameters model, DataViewSchema trainSchema)
+            => new BinaryPredictionTransformer<TModelParameters>(Host, model, trainSchema, FeatureColumn.Name);
 
         /// <summary>
         /// Continues the training of a <see cref="SdcaBinaryTrainer"/> using an initial predictor and returns a <see cref="BinaryPredictionTransformer"/>.
         /// </summary>
-        public BinaryPredictionTransformer<TScalarPredictor> Fit(IDataView trainData, IPredictor initialPredictor)
+        public BinaryPredictionTransformer<TModelParameters> Fit(IDataView trainData, IPredictor initialPredictor = null)
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.StandardLearners/Standard/SdcaBinary.cs]: >= null [](start = 114, length = 6)

don't change it please.

File: test/Microsoft.ML.Tests/Scenarios/Api/CookbookSamples/CookbookSamplesDynamicApi.cs [hunk 4]
```diff
@@ -108,7 +108,7 @@ private void TrainRegression(string trainDataPath, string testDataPath, string m
                 // once so adding a caching step before it is not helpful.
                 .AppendCacheCheckpoint(mlContext)
                 // Add the SDCA regression trainer.
-                .Append(mlContext.Regression.Trainers.StochasticDualCoordinateAscent(labelColumn: "Target", featureColumn: "FeatureVector"));
+                .Append(mlContext.Regression.Trainers.StochasticDualCoordinateAscent(labelColumnName: "Target", featureColumnName: "FeatureVector"));
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Tests/Scenarios/Api/CookbookSamples/CookbookSamplesDynamicApi.cs]: >StochasticDualCoordinateAscent [](start = 54, length = 30)

update [https://github.com/dotnet/machinelearning/blob/master/docs/code/MlNetCookBook.md](https://github.com/dotnet/machinelearning/blob/master/docs/code/MlNetCookBook.md)

File: test/Microsoft.ML.Tests/TrainerEstimators/TrainerEstimators.cs [hunk 5]
```diff
@@ -82,20 +83,62 @@ public void KMeansEstimator()
...
+                var pipeWithTrainer = pipe.Append(trainer);
+                TestEstimatorCore(pipeWithTrainer, dataView);
+
+                var transformedDataView = pipe.Fit(dataView).Transform(dataView);
+                var model = trainer.Fit(transformedDataView);
+                trainer.Fit(transformedDataView, model.Model);
+                TestEstimatorCore(pipe, dataView);
+
+                var result = model.Transform(transformedDataView);
+                var metrics = ML.BinaryClassification.EvaluateNonCalibrated(result);
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Tests/TrainerEstimators/TrainerEstimators.cs]: >EvaluateNonCalibrated [](start = 54, length = 21)

Doesn't this one produce calibrated model?

File: src/Microsoft.ML.StandardLearners/Standard/SdcaBinary.cs (hunk 3 above)
Comment (wschin) [reply src/Microsoft.ML.StandardLearners/Standard/SdcaBinary.cs]: TrainTransformer is typed to TModelParameters, so we need this change.

---
In reply to: [257348893](https://github.com/dotnet/machinelearning/pull/2564#discussion_r257348893) [](ancestors = 257348893)

Comment (Ivanidzo4ka) [reply src/Microsoft.ML.StandardLearners/Standard/SdcaBinary.cs]: If you look on my selection I was talking about `IPredictor initialPredictor = null` `=null` part

---
In reply to: [257351775](https://github.com/dotnet/machinelearning/pull/2564#discussion_r257351775) [](ancestors = 257351775,257348893)

Comment (wschin) [reply src/Microsoft.ML.StandardLearners/Standard/SdcaBinary.cs]: Ah. My bad.

---
In reply to: [257352673](https://github.com/dotnet/machinelearning/pull/2564#discussion_r257352673) [](ancestors = 257352673,257351775,257348893)

File: test/Microsoft.ML.Tests/TrainerEstimators/TrainerEstimators.cs (hunk 5 above)
Comment (wschin) [reply test/Microsoft.ML.Tests/TrainerEstimators/TrainerEstimators.cs]: Right. Will test logloss too.

---
In reply to: [257351073](https://github.com/dotnet/machinelearning/pull/2564#discussion_r257351073) [](ancestors = 257351073)
